exchanges/gateio/gateio_websocket.go

```diff
@@ -764,6 +733,146 @@ func singleSymbolChannel(name string) bool {
 	return false
 }
 
+// ValidateSubscriptions implements subscription.SubscriptionsValidator interface
+func (g *Gateio) ValidateSubscriptions(l subscription.List) error {
+	orderbookGuard := map[key.PairAsset]string{}
+	for _, s := range l {
+		n := channelName(s)
+		if !isRestrictedOrderbookChannel(n) {
+			continue
+		}
+		for _, p := range s.Pairs {
+			k := key.PairAsset{Base: p.Base.Item, Quote: p.Quote.Item, Asset: s.Asset}
+			loadedChanName, ok := orderbookGuard[k]
+			if !ok {
+				orderbookGuard[k] = n
+				continue
+			}
+			if loadedChanName != n {
+				return fmt.Errorf("%w for `%s %s` between `%s` and `%s`, please enable only one type", subscription.ErrExclusiveSubscription, k.Pair(), k.Asset, loadedChanName, n)
```

Comment:
Am I right in thinking that this is fixing this error:
`error: malformed orderbook data`
If so, that makes sense why we need this function in general, we can't trust them to check for us, and we need to be checking against all subscriptions.

Reply:
That error was to do with incremental updates and the fact that zero values came through which is needed because because it updates the sequence numbers. When multiple book channels are subscribed to though, this throws the internal book completely out.